doc/man3/BIO_sendmmsg.pod

```diff
@@ -197,6 +197,10 @@ transient error, many of which are system specific.
 
 =back
 
+Third parties implementing custom BIOs supporting the BIO_sendmmsg() or
+BIO_recvmmsg() methods should note that it is a required part of the API
+contract that an error is always raised when either of these functions return 0.
```

Comment:
This seems slightly problematic. What happens if they do not raise an error?

Reply:
This problem is intrinsic to the design of the ERR API, as was discussed at the committers/OTC F2F.

The problem is that if an implementation of BIO_sendmmsg doesn't honour this API contract, then a caller which calls BIO_sendmmsg and has it return 0 doesn't know if the ERR on the top of the stack is from BIO_sendmmsg or just left over junk from some previous call to some other API. It is possible for a caller to work around this by always clearing the ERR stack before calling an API, and this would work for a faulty BIO_sendmmsg implementation. But it is vastly preferable for implementations to simply honour this requirement.

This is basically an example instance of the "prime directive" as proposed in the document/slides I gave at the F2F.

I will take this opportunity to note something else, though: in the case of BIO_sendmmsg/BIO_recvmmsg, we are using this for "normative" (as opposed to diagnostic) error handling, meaning error handling that affects control flow and is expected to happen in "normal" circumstances.

We have previously expressed a policy, now noted in our documentation, that ERRs being raised is not part of our API contract and should not be depended on by applications, *except* where documentation notes otherwise. But the general understanding seems to be that ERR is intended for diagnostics and not control flow.

The deeper problem this leads to is that we don't actually have any standard API in OpenSSL for expressing error conditions needed to allow applications to make control flow decisions. The closest we have is actually SSL_get_error, which is highly specific. Other than that we end up having to document use of ERR as part of an API contract for specific functions when there's no other way. It's also very likely a lot of applications end up depending on ERRs being raised which we *don't* document as part of our API contract, making the recent addition of the statement that we don't guarantee anything unless explicitly specified to our documentation a relatively radical move IMO.

The proposal I gave at the F2F is to make ERR at least sound for the purposes of usage with control flow. Another alternative which I considered proposing, but left out for the sake of brevity, is a POSIX style thread-local `errno` variable which stores a single error code integer (`OSSL_get_error_code` or something).  This would be explicitly designed for control flow handling and would not give out surplus (diagnostic) information. The prime directive would still apply however (it must be completely predictable from documentation whether the variable is updated by a given function when it returns 0), so actually the same situation as is being discussed here.

Anyway, there are only a few options for this immediate issue:

- Keep the current design, which uses ERR and basically adopts the prime directive
  for ERR usage for this API specifically to make it easier to use it sanely. As a fallback in the face of a buggy custom BIO implementation an application can always clear the ERR stack before calling the API.
- Come up with some custom way of outputting error information from BIO_sendmmsg/BIO_recvmmsg. e.g. an existing ERR packed error code in an output parameter. Not having to deal with ERR might be nice, but the problem is that this is basically doing things completely differently to any other OpenSSL API. I am not convinced the benefits here outweigh the disadvantages of the idiosyncrasy of this approach.

@mattcaswell

Reply:
Thank you for the explanation. IMO the idea to have the OSSL_get_error_code() does not sound that bad. We could adopt it gradually. I do not think we want to introduce it in 3.2 though.